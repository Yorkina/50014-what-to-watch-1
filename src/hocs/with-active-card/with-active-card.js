import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';


const DELAY_TIME = 1000;

export const withActiveCard = (Component) => {
  class WithActiveCard extends PureComponent {
    constructor(props) {
      super(props);

      this._timerId = 0;

      this.state = {
        isPlaying: false,
      };

      this._handleMouseEnter = this._handleMouseEnter.bind(this);
      this._handleMouseLeave = this._handleMouseLeave.bind(this);
      this._handleTitleClick = this._handleTitleClick.bind(this);
    }

    _handleMouseEnter() {
      this._timerId = setTimeout(() => {
        this.setState({
          isPlaying: true,
        });
      }, DELAY_TIME);
    }

    _handleTitleClick() {
      return;
    }

    _handleMouseLeave() {
      clearTimeout(this._timerId);

      this.setState({
        isPlaying: false,
      });
    }

    componentWillUnmount() {
      clearTimeout(this._timerId);
    }

    render() {
      const {isPlaying} = this.state;
      const {card} = this.props;

      return <Component
        card={card}
        isPlaying={isPlaying}
        onTitleClick={this._handleTitleClick}
        onMouseEnter={this._handleMouseEnter}
        onMouseLeave={this._handleMouseLeave}
      />;
    }
  }

  WithActiveCard.propTypes = {
    card: PropTypes.shape({
      href: PropTypes.string.isRequred,
      preview: PropTypes.string.isRequred,
      src: PropTypes.string.isRequred,
      title: PropTypes.string.isRequred,
    })
  };

  return WithActiveCard;
}

export default withActiveCard;
