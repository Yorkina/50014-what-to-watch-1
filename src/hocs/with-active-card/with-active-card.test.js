import React from 'react';
import renderer from 'react-test-renderer';

import {withActiveCard} from './with-active-card';


const MockComponent = () => <div />;
const MockComponentWrapped = withActiveCard(MockComponent);

describe(`WithActiveCard`, () => {
  it(`should render`, () => {
    const props = {
      card: {
        href: `fantastic-beasts-the-crimes-of-grindelwald.html`,
        previewImageSrc: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        name: `Fantastic Beasts: The Crimes of Grindelwald`,
      }
    };

    const tree = renderer.create(<MockComponentWrapped {...props}/>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
