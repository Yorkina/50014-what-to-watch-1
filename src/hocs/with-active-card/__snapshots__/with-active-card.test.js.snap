// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithActiveCard should render 1`] = `<div />`;
